misc/synchronization.md

```diff
@@ -0,0 +1,368 @@
+- the destructor notifies that no new tasks will be scheduled on any of the tile,
+- `done(tile_index)` notifies that no new task will be scheduled on the tile with the given index
+(After this call any use of `operator()` or `read()` with the same tile index will return an error.),
+- `doneWrite(tile_index)` notifies that no new read-write task will be scheduled on the tile with the given index
+(After this call any use of `operator()` with the same tile index will return an error.).
+
+It has to be noted that read-only task of the parent matrix and of the View can be performed at the same time
+if no read-write tasks has been scheduled in between (see [examples](#view-examples)).
+
+### View Examples
+
+(Note: this examples are used to illustrate how dependencies are determined,
+however they cannot be compiled since they are simplified to improve readability.
+(e.g. constructors and template parameters are omitted)).
+
+- Example 1: Basic view usage.
+```
+  Matrix m;
+
+  hpx::dataflow(Task1, m({0, 0})));
+  hpx::dataflow(Task2, m({0, 0})));  // Depends on Task1.
+  hpx::dataflow(Task3, m({0, 1})));  // Different tile. No dependency on Task1 or Task2.
+
+  {
+    MatrixView mv(UpLo::General, m);
+    hpx::dataflow(Task4, mv({0, 0})));  // Depends on Task2.
+    hpx::dataflow(Task5, mv({0, 1})));  // Depends on Task3.
+  }
+
+  hpx::dataflow(Task6, m({0, 0})));  // Depends on Task4.
+  hpx::dataflow(Task7, m({0, 1})));  // Depends on Task5.
+
+// Resulting dependency graph:
+// Task1 - Task2 - Task4 ~ Task6
+//
+// Task3 - Task5 ~ Task7
+//
+// Note the that the ~ dependencies include the call to the destructor of mv.
+```
+- Example 2: Basic Usage with `done()`.
+```
+  Matrix m;
+
+  hpx::dataflow(Task1, m({0, 0})));
+  hpx::dataflow(Task2, m({0, 0})));  // Depends on Task1.
+  hpx::dataflow(Task3, m({0, 1})));  // Different tile. No dependency on Task1 or Task2.
+
+  MatrixView mv(UpLo::General, m);
+
+  hpx::dataflow(Task4, m({0, 0})));  // Depends on done notification and Task6
+  hpx::dataflow(Task5, m({0, 1})));  // Depends on done notification and Task7
+  hpx::dataflow(Task6, mv({0, 0})));  // Depends on Task2.
+  hpx::dataflow(Task7, mv({0, 1})));  // Depends on Task3.
+
+  mv.done({0, 0});
+  mv.done({0, 1});
+
+// Resulting dependency graph:
+// Task1 - Task2 - Task6 ~ Task4
+//
+// Task3 - Task7 ~ Task5
+//
+// Note the that the ~ dependencies include the dependency on the call to done().
+```
+- Example 3: Read cuncurrency.
+```
+  Matrix m;
+
+  hpx::dataflow(Task1, m({0, 0})));
+  hpx::dataflow(Task2, m.read({0, 0})));  // Depends on Task1
+
+  MatrixView mv(UpLo::General, m);
+
+  hpx::dataflow(Task3, m.read({0, 0})));  // Depends on doneWrite notification and Task5
+  hpx::dataflow(Task4, mv.read({0, 0})));  // Depends on Task2.
+  hpx::dataflow(Task5, mv({0, 0})));  // Depends on Task2 and Task4.
+  hpx::dataflow(Task6, mv.read({0, 0})));  // Depends on Task5.
+
+  mv.doneWrite({0, 0});
```

Comment:
I'm not sure if putting this call in the "middle" of the dag construction can be useful or misleading.

Reply:
Yes, the positions of done and doneWrite are a bit messed up.
I would put them as soon as they can be called, i.e.:

- doneWrite after creation of task5
- done after task6

What do you think?